### DISCUSSION TITLE
Nak packets from host to implement flow control

### DISCUSSION THREAD

**Discussion by taimoor-naveed** [hathach/tinyusb#1366] [Q&A]
I'm using tinyusb to implement a generic hid device. Everything seems to work fine and I can exchange data between the host and the MCU.
Now I'd like to NAK packets coming from the host until the previous packet has been processed. It seems possible when using the stack provided by ST by delaying the call to the function which prepares for the next transfer. Until you invoke USBD_LL_PrepareReceive(), all further incoming packets are NAK'd
How can I achieve the same functionality with tinyusb?
P. S. I'm using the hid generic inout example as base woth freertos

**Comment by hathach** [comment]
NAK is default behavior if there is no data, you don't have to do anything.

**Reply by taimoor-naveed** [reply]
Hi @hathach, Thank you for your reply. Let me elaborate a bit on what I mean.
Currently, whenever the function, ```tud_hid_set_report_cb``` is called, I push the data into a queue.
At the same time, there is another task that is waiting on the queue. As soon as the data is pushed, the task wakes up and then starts to process the data. The time it takes to process the data is variable. It can take up to 10ms in the worst case. how can i make sure that the host doesn't send any more data when the task is already processing data. 
I can just increase the capacity of the queue that holds the data but that too is not a viable solution for me because even if i set the queue size to a big value, it will still fill up as the consumption of data is slower than the host's speed to send data
The current polling interval I'm using is set to 1ms. I can't also increase the poll interval in the usb descriptor because most of the time, the data is processed under 1ms. Only sometimes it takes up to 10ms. But the application is designed in such a way that I don't know when and how often the worst case scenario will occur

**Reply by hathach** [reply]
as mentioned above, if you don't make any transfer. it will response with NAK. For reproducing, could you tell
1. which mcu/board you are running
2. which example you are running, it should be one of the stock example, your case is probably hid_generic_inout
3. how do you send data from host to device ? and what issue it is causing, missing data ? If so you could post your script and device output here. Please attach txt file if it is too long.
